### PR TITLE
Add native registry login tests for EKS, AKS and GKE

### PR DESCRIPTION
**Pull request by darkowlzz**
It introduces a new package `tftestenv`. It is similar to [testenv](https://pkg.go.dev/github.com/fluxcd/pkg/runtime@v0.16.2/testenv) but uses
terraform to create test environment. It provides helpers to run terraform
apply and destroy in the test context.

`tftestenv` is used to create tests similar to the controller tests but with
cloud provider managed cluster provisioned using terraform. The test
setup is written similar to the usual controller test suite_test.go,
providing flexibility in the test setup and execution. The structure tries
to keep the infrastructure code separate from the test code such that
the infrastructure provisioning tool can be easily replaced, without the
need to change the actual tests.

The go tests are also written similar to the usual controller tests,
with a kube client from the tftestenv environment global variable.

Terraform is automatically downloaded using the terraform-exec package.
The test container images are generated using the go-containerregistry
package and pushed to the test container image repository.

`tests/integration/README.md` contains more details about the test setup and workflow.

The terraform configurations for EKS is based on https://github.com/terraform-aws-modules/terraform-aws-eks/tree/6bb73bbdd469797cb067e9aa22ce2f31a7fb22be/examples/self_managed_node_group .

The terraform configurations for AKS is based on https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster .

The terraform configurations for GKE is based on the official GKE terraform provider examples https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/container_cluster with [Terraform Kubernetes Engine Auth Module](https://registry.terraform.io/modules/terraform-google-modules/kubernetes-engine/google/21.1.0/submodules/auth) to be able to easily retrieve the kubeconfig. Both GCR and Artifact Registry are tested.

Average test duration including provisioning and cleanup time:
- EKS - 25-30 minutes
- AKS - ~13 minutes
- GKE - ~9 minutes

We may be able to optimize some configurations to bring down the time of EKS and AKS in the future.

Part of  #264.

### CODE REVIEW

**Comment by darkowlzz** [comment]
Posting some results to provide an idea of what it looks like when run.

With `.env`

```sh
# AWS
export AWS_ACCESS_KEY_ID=<redacted>
export AWS_SECRET_ACCESS_KEY=<redacted>
export AWS_REGION=us-east-2

# Azure
export TF_VAR_azure_location=eastus

# GCP
export TF_VAR_gcp_project_id=<redacted>
# export TF_VAR_gcp_region=us-central1
# export TF_VAR_gcp_zone=us-central1-c
# Leave GCR region empty to use gcr.io. Else set it to `us`, `eu` or `asia`.
# export TF_VAR_gcr_region=
```

and `az login` and `gcloud auth login` being already run:

aws:

```console
$ make test-aws 
make test PROVIDER_ARG="-provider aws"
make[1]: Entering directory '/home/go/src/github.com/fluxcd/image-reflector-controller/tests/integration'
cp kustomization.yaml build/flux
cd build/flux && kustomize edit set image fluxcd/image-reflector-controller=fluxcd/image-reflector-controller
kustomize build build/flux > build/flux.yaml
go test -timeout 30m -v ./...  -provider aws
2022/06/23 02:53:59 Terraform binary:  /usr/bin/terraform
2022/06/23 02:53:59 Init Terraform
2022/06/23 02:54:05 Checking for an empty Terraform state
2022/06/23 02:54:07 Applying Terraform
2022/06/23 03:08:34 pushing test image z.dkr.ecr.us-east-2.amazonaws.com/flux-test-repo-peaceful-ghost:v0.1.0
2022/06/23 03:08:38 pushing test image z.dkr.ecr.us-east-2.amazonaws.com/flux-test-repo-peaceful-ghost:v0.1.2
2022/06/23 03:08:41 pushing test image z.dkr.ecr.us-east-2.amazonaws.com/flux-test-repo-peaceful-ghost:v0.1.3
2022/06/23 03:08:44 pushing test image z.dkr.ecr.us-east-2.amazonaws.com/flux-test-repo-peaceful-ghost:v0.1.4
2022/06/23 03:08:47 Installing flux
=== RUN   TestImageRepositoryScan
=== RUN   TestImageRepositoryScan/ecr
--- PASS: TestImageRepositoryScan (2.01s)
    --- PASS: TestImageRepositoryScan/ecr (2.01s)
PASS
2022/06/23 03:09:20 Destroying environment...
ok      github.com/fluxcd/image-reflector-controller/tests/integration  1592.611s
make[1]: Leaving directory '/home/go/src/github.com/fluxcd/image-reflector-controller/tests/integration'
```

azure:

```console
$ make test-azure 
make test PROVIDER_ARG="-provider azure"
make[1]: Entering directory '/home/go/src/github.com/fluxcd/image-reflector-controller/tests/integration'
cp kustomization.yaml build/flux
cd build/flux && kustomize edit set image fluxcd/image-reflector-controller=fluxcd/image-reflector-controller
kustomize build build/flux > build/flux.yaml
go test -timeout 30m -v ./...  -provider azure
2022/06/23 03:23:50 Terraform binary:  /usr/bin/terraform
2022/06/23 03:23:50 Init Terraform
2022/06/23 03:23:52 Checking for an empty Terraform state
2022/06/23 03:23:52 Applying Terraform
2022/06/23 03:29:34 pushing test image fluxtestrepomovingotick.azurecr.io/vynz8:v0.1.0
2022/06/23 03:29:38 pushing test image fluxtestrepomovingotick.azurecr.io/vynz8:v0.1.2
2022/06/23 03:29:40 pushing test image fluxtestrepomovingotick.azurecr.io/vynz8:v0.1.3
2022/06/23 03:29:42 pushing test image fluxtestrepomovingotick.azurecr.io/vynz8:v0.1.4
2022/06/23 03:29:45 Installing flux
=== RUN   TestImageRepositoryScan
=== RUN   TestImageRepositoryScan/acr
--- PASS: TestImageRepositoryScan (3.36s)
    --- PASS: TestImageRepositoryScan/acr (3.36s)
PASS
2022/06/23 03:30:16 Destroying environment...
ok      github.com/fluxcd/image-reflector-controller/tests/integration  749.433s
make[1]: Leaving directory '/home/go/src/github.com/fluxcd/image-reflector-controller/tests/integration'
```

gcp:

```console
$ make test-gcp 
make test PROVIDER_ARG="-provider gcp"
make[1]: Entering directory '/home/go/src/github.com/fluxcd/image-reflector-controller/tests/integration'
cp kustomization.yaml build/flux
cd build/flux && kustomize edit set image fluxcd/image-reflector-controller=fluxcd/image-reflector-controller
kustomize build build/flux > build/flux.yaml
go test -timeout 30m -v ./...  -provider gcp
2022/06/23 03:39:18 Terraform binary:  /usr/bin/terraform
2022/06/23 03:39:18 Init Terraform
2022/06/23 03:39:22 Checking for an empty Terraform state
2022/06/23 03:39:23 Applying Terraform
2022/06/23 03:44:37 pushing test image gcr.io/z/ik9ar:v0.1.0
2022/06/23 03:44:46 pushing test image gcr.io/z/ik9ar:v0.1.2
2022/06/23 03:44:54 pushing test image gcr.io/z/ik9ar:v0.1.3
2022/06/23 03:45:02 pushing test image gcr.io/z/ik9ar:v0.1.4
2022/06/23 03:45:10 pushing test image us-central1-docker.pkg.dev/z/flux-test-repo-helpful-kangaroo/8y697:v0.1.0
2022/06/23 03:45:19 pushing test image us-central1-docker.pkg.dev/z/flux-test-repo-helpful-kangaroo/8y697:v0.1.2
2022/06/23 03:45:26 pushing test image us-central1-docker.pkg.dev/z/flux-test-repo-helpful-kangaroo/8y697:v0.1.3
2022/06/23 03:45:32 pushing test image us-central1-docker.pkg.dev/z/flux-test-repo-helpful-kangaroo/8y697:v0.1.4
2022/06/23 03:45:38 Installing flux
=== RUN   TestImageRepositoryScan
=== RUN   TestImageRepositoryScan/gcr
=== RUN   TestImageRepositoryScan/artifact_registry
--- PASS: TestImageRepositoryScan (9.11s)
    --- PASS: TestImageRepositoryScan/gcr (8.29s)
    --- PASS: TestImageRepositoryScan/artifact_registry (0.82s)
PASS
2022/06/23 03:46:17 Destroying environment...
ok      github.com/fluxcd/image-reflector-controller/tests/integration  568.476s
make[1]: Leaving directory '/home/go/src/github.com/fluxcd/image-reflector-controller/tests/integration'
```

File: tests/integration/azure_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	if !ok || kubeconfigYaml == "" {
+		return fmt.Errorf("failed to obtain kubeconfig from tf output")
+	}
+
+	f, err := os.Create(kcPath)
+	if err != nil {
+		return err
+	}
+	_, err = fmt.Fprint(f, kubeconfigYaml)
+	return err
```
Comment (pjbgf) [inline tests/integration/azure_test.go:42]: We need to close the file handle `f.Close()` and check for errors as well:

```golang
	if err := f.Close(); err != nil {
            return err
	}
```

File: tests/integration/gcp_test.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	if !ok || kubeconfigYaml == "" {
+		return fmt.Errorf("failed to obtain kubeconfig from tf output")
+	}
+
+	f, err := os.Create(kcPath)
+	if err != nil {
+		return err
+	}
+	_, err = fmt.Fprint(f, kubeconfigYaml)
+	return err
```
Comment (pjbgf) [inline tests/integration/gcp_test.go:42]: Same here, we need to close the file handle `f.Close()` and check for errors as well:

```golang
	if err := f.Close(); err != nil {
            return err
	}
```

File: tests/integration/gcp_test.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	// with a new repository name.
+	repoURL := output["gcr_repository_url"].Value.(string)
+	if err := tftestenv.RunCommand(ctx, "./",
+		fmt.Sprintf("gcloud auth configure-docker %s", repoURL),
+		tftestenv.RunCommandOptions{},
+	); err != nil {
+		return nil, err
+	}
+
+	// NOTE: Artifact Registry calls a registry a repository. A repository can
```
Comment (pjbgf) [inline tests/integration/gcp_test.go:58]: Maybe?
```suggestion
	// NOTE: Artifact Registry calls a registry "repository". A repository can
```

File: tests/integration/README.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
+# Cloud Provider Integration Tests
+
+## Requirements
+
+### Amazon Web Services
+
+- AWS account with access key ID and secret access key with permissions to
+    create EKS cluster and ECR repository.
+- AWS CLI, need not be configured with the AWS account.
```
Comment (pjbgf) [inline tests/integration/README.md:9]: Do you mean:
a) It **does not need** to be configured with the AWS account
or
b) It **must not** be configured with the AWS account?

Comment (pjbgf) [reply tests/integration/README.md:9]: I would then rephrase it to something like:

```suggestion
- AWS CLI, which does not need to be configured with the AWS account above.
```

**Comment by phillebaba** [comment]
This looks really good but I think some of these things would be good to share with the [Azure E2E tests](https://github.com/fluxcd/flux2/tree/main/tests/azure). Maybe we should move some of the packages to pkg instead?

**Comment by darkowlzz** [comment]
> This looks really good but I think some of these things would be good to share with the [Azure E2E tests](https://github.com/fluxcd/flux2/tree/main/tests/azure). Maybe we should move some of the packages to pkg instead?

@phillebaba yes, that's the plan after getting these changes reviewed, accepted and merged here initially.

**Review by phillebaba — APPROVED** [review]
Small NIT but otherwise great!

**Comment by darkowlzz** [comment]
Rebased and verified that it works with go 1.18.

**Review by hiddeco — APPROVED** [review]
